### DISCUSSION TITLE
How to export rules severity, pillar and pillar subcategory

### DISCUSSION THREAD

**Discussion by jsporbert** [Azure/PSRule.Rules.Azure#1986] [Q&A] [answered]
In order to classify rules, I'd like to build an excel file in which I can classify rules by WAF pillar or severity..CSV does not include this columns and I can't find a way to do it using powershell :( . Any hint or is it a new feature request ? :) 

TIA

**Comment by BernieWhite** [comment, marked as answer]
The list of columns in the built-in CSV output method is not configurable. There is a issue over here to make this more customizable microsoft/PSRule#1165.

For results, you could call `Invoke-PSRule` and `Select-Object` to expose the specific columns you are looking for.

```powershell
Invoke-PSRule -f . -Format File -Outcome Fail | Select-Object RuleName, @{ Name = "Severity"; Expression = { $_.Info.Annotations.severity }}, @{ Name = "Pillar"; Expression = { $_.Info.Annotations.pillar }} | ConvertTo-Csv | Set-Content -Path out.csv
```

For exporting the pillar and severity data for merging in you could use `Get-PSRule` to export details for each rule:

```powershell
 Get-PSRule -Module PSRule.Rules.Azure | Select-Object RuleName, @{ Name = "Severity"; Expression = { $_.Info.Annotations.severity }}, @{ Name = "Pillar"; Expression = { $_.Info.Annotations.pillar }}
```

---

I hope that helps.

**Reply by jsporbert** [reply]
Thanks a lot ! It definitely helps, I missed the property where it was stored :)

**Comment by mschwrdtnr** [comment]
Hi @BernieWhite, i created a small powershell script which adds Pillar, Severity and Category to the csv export. Is there any way i can contribute it in this project?

**Reply by BernieWhite** [reply]
@mschwrdtnr I think adding it to samples or docs would make the most sense currently.

As you have worked out, there is a plan to add support more broadly in PSRule. Happy to have suggestions on specifics in the microsoft/PSRule repo.

Samples can be added here: https://github.com/microsoft/PSRule-samples/tree/main/samples/azure

Docs can be added as a new customization page here: https://github.com/Azure/PSRule.Rules.Azure/tree/main/docs/customization

**Comment by azarchitect2024** [comment]
I have found an alternate approach to get this done. Here are the steps: 

Step 1: As @BernieWhite  mentioned run the below command to create a master reference sheet. This sheet carries the Azure Rule name, Severity and the WAF Pillar. Save this CSV as excel and this is your baseline for WAF assessment.

Step 2: Run the following command to get the analysis report.
Invoke-PSRule -InputPath $path -Module 'PSRule.Rules.Azure' -OutputFormat Csv -OutputPath $path\export1.csv

** Here you can declare the $path variable with your preferable location where the JSON is living.

Save the CSV as a excel file to help in further Step 3.

Step 3: Run a excel vlookup on the excel from step 2 with excel from step 1 as a reference (use the Azure Rule Name as a reference). Excel will do the rest and you get an amazing report covering all that you need. I am attaching screenshot from my demo azure lab environment so you see the fields. 

Once again this is not a 100% automated solution , but does the job :)
<img width="921" alt="sample" src="https://github.com/Azure/PSRule.Rules.Azure/assets/159544445/5d2abf56-1e38-4c8d-b810-53bcf077f89c">

<img width="278" alt="sample2" src="https://github.com/Azure/PSRule.Rules.Azure/assets/159544445/dd033095-6e72-4968-96b0-e9b48dafd605">

**Reply by BernieWhite** [reply]
Nice suggestion @azarchitect2024
